host/xtest/regression_4000.c

```diff
@@ -6192,3 +6192,498 @@ static void xtest_tee_test_4016_ed25519(ADBG_Case_t *c)
+
+	if (bs->auth_enc) {
+		if (bs->decrypt)
+			res = ta_crypt_cmd_ae_decrypt_final(c, bs->s, bs->oph,
+							    src, src_len,
+							    NULL, &dlen,
+							    bs->tag,
+							    sizeof(bs->tag));
+		else
+			res = ta_crypt_cmd_ae_encrypt_final(c, bs->s, bs->oph,
+							    src, src_len,
+							    NULL, &dlen, NULL,
+							    &tlen);
+		if (!res && bs->decrypt)
+			return ADBG_EXPECT_COMPARE_UNSIGNED(c, dlen, ==,
+							   src_len);
+		if (!ADBG_EXPECT_TEEC_RESULT(c, TEEC_ERROR_SHORT_BUFFER, res) ||
+		    !ADBG_EXPECT_COMPARE_UNSIGNED(c, bs->text_size -
+                                                  bs->out_count, ==, dlen))
+			return false;
+		if (!bs->decrypt &&
+		    !ADBG_EXPECT_COMPARE_UNSIGNED(c, sizeof(bs->tag), ==, tlen))
+			return false;
+		if (bs->decrypt)
+			res = ta_crypt_cmd_ae_decrypt_final(c, bs->s, bs->oph,
+							    src, src_len,
+							    bs->tmp_text, &dlen,
+							    bs->tag,
+							    sizeof(bs->tag));
+		else
+			res = ta_crypt_cmd_ae_encrypt_final(c, bs->s, bs->oph,
+							    src, src_len,
+							    bs->tmp_text, &dlen,
+							    bs->tmp_tag, &tlen);
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c, res) ||
+		    !ADBG_EXPECT_BUFFER(c, reference,
+					bs->text_size - bs->out_count,
+					bs->tmp_text, dlen))
+			return false;
+		if (!bs->decrypt &&
+		    !ADBG_EXPECT_BUFFER(c, bs->tag, sizeof(bs->tag),
+					bs->tmp_tag, tlen))
+			return false;
+	} else {
+		res = ta_crypt_cmd_cipher_do_final(c, bs->s, bs->oph,
+						   src, src_len, NULL, &dlen);
+		if (!res)
+			return ADBG_EXPECT_COMPARE_UNSIGNED(c, dlen, ==,
+							    src_len);
+		if (!ADBG_EXPECT_TEEC_RESULT(c, TEEC_ERROR_SHORT_BUFFER, res) ||
+		    !ADBG_EXPECT_COMPARE_UNSIGNED(c,
+						  bs->text_size -
+						  bs->out_count, ==, dlen))
+			return false;
+
+		res = ta_crypt_cmd_cipher_do_final(c, bs->s, bs->oph,
+						   src, src_len,
+						   bs->tmp_text, &dlen);
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c, res) ||
+		    !ADBG_EXPECT_BUFFER(c, reference,
+					bs->text_size - bs->out_count,
+					bs->tmp_text, dlen))
+			return false;
+	}
+
+	return true;
+}
+
+static bool process_text_4017(ADBG_Case_t *c, struct buf_state *bs,
+			      size_t initial_count, size_t middle_count,
+			      size_t middle_increment)
+{
+	TEEC_Result res = TEEC_SUCCESS;
+	size_t n = 0;
+
+	bs->in_count = 0;
+	bs->out_count = 0;
+	bs->buf_count = 0;
+
+	if (bs->auth_enc)
+		res = ta_crypt_cmd_ae_init(c, bs->s, bs->oph, bs->iv,
+					   bs->iv_len, sizeof(bs->tag), 0,
+					   bs->text_size);
+	else
+		res = ta_crypt_cmd_cipher_init(c, bs->s, bs->oph, bs->iv,
+					       bs->iv_len);
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, res))
+		return false;
+
+	if (initial_count &&
+	    !ADBG_EXPECT_TRUE(c, process_bytes_4017(c, bs, initial_count)))
+		return false;
+
+	for (n = 0; n < middle_count && n + initial_count < bs->text_size;
+	     n += middle_increment)
+		if (!ADBG_EXPECT_TRUE(c, process_bytes_4017(c, bs, 1)))
+			return false;
```

Comment:
this looks strange to me. I think it works because `middle_increment` is always 1, as stated in the inline comment at line 6561:
```
		 * If level < 12 test with middle_count of 32 else 0.
		 * (middle_count - n) bytes are processed one by one.
```

Unless you plan some day to change this `middle_increment` in the future, I think it would more better to remove `middle_increment` and use `1` straight. Otherwise, I suggest to change line 6396:
```patch
-		if (!ADBG_EXPECT_TRUE(c, process_bytes_4017(c, bs, 1)))
+		if (!ADBG_EXPECT_TRUE(c, process_bytes_4017(c, bs, middle_increment)))
```
and line 6562:
```patch
 		 * If level < 12 test with middle_count of 32 else 0.
-		 * (middle_count - n) bytes are processed one by one.
+		 * (middle_count - n) bytes are processed by middle_increment
+		 * bytes bursts.
```

If I understood that correctly of course :)

Reply:
I'll drop the `middle_increment`.